sdks/java/io/hive/src/main/java/org/apache/beam/sdk/io/hive/HiveIO.java

```diff
@@ -0,0 +1,474 @@
+    @Nullable abstract String getMetastoreUri();
+    @Nullable abstract String getDatabase();
+    @Nullable abstract String getTable();
+    @Nullable abstract Map getFilter();
+    abstract long getBatchSize();
+
+    abstract Builder toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setMetastoreUri(String metastoreUri);
+      abstract Builder setDatabase(String database);
+      abstract Builder setTable(String table);
+      abstract Builder setFilter(Map partition);
+      abstract Builder setBatchSize(long batchSize);
+      abstract Write build();
+    }
+
+    public Write withMetastoreUri(String metastoreUri) {
+      return toBuilder().setMetastoreUri(metastoreUri).build();
+    }
+
+    public Write withDatabase(String database) {
+      return toBuilder().setDatabase(database).build();
+    }
+
+    public Write withTable(String table) {
+      return toBuilder().setTable(table).build();
+    }
+
+    public Write withFilter(Map filter) {
+      return toBuilder().setFilter(filter).build();
+    }
+
+    public Write withBatchSize(long batchSize) {
+      return toBuilder().setBatchSize(batchSize).build();
+    }
+
+    @Override
+    public PDone expand(PCollection<HCatRecord> input) {
+      input.apply(ParDo.of(new WriteFn(this)));
+      return PDone.in(input.getPipeline());
+    }
+
+    @Override
+    public void validate(PipelineOptions options) {
+      checkNotNull(getMetastoreUri(), "metastoreUri");
+      checkNotNull(getTable(), "database");
+    }
+
+    private static class WriteFn extends DoFn<HCatRecord, Void> {
+      private final Write spec;
+      private WriterContext writerContext;
+      private HCatWriter slaveWriter;
+      private HCatWriter masterWriter;
+      private List<HCatRecord> hCatRecordsBatch;
+
+      public WriteFn(Write spec) {
+        this.spec = spec;
+      }
+
+      @Override
+      public void populateDisplayData(
+          org.apache.beam.sdk.transforms.display.DisplayData.Builder builder) {
+        super.populateDisplayData(builder);
+        builder.add(DisplayData.item("database", spec.getDatabase()));
+        builder.add(DisplayData.item("table", spec.getTable()));
+        builder.add(DisplayData.item("filter", (spec.getFilter() != null
+            ? spec.getFilter().toString() : "NULL")));
+        builder.add(DisplayData.item("metastoreUri", spec.getMetastoreUri()));
+        builder.add(DisplayData.item("batchSize", spec.getBatchSize()));
+      }
+
+      @Setup
+      public void initiateWrite() throws HCatException {
+        try {
+          //initiate writerContext, masterWriter & slaveWriter
+          //to be used for writing into Hive
+          writerContext = getWriterContext();
+          slaveWriter = DataTransferFactory.getHCatWriter(writerContext);
+        } catch (HCatException e) {
+          LOG.error("Exception in initiateWriterContext", e);
```

Comment:
This is an unnecessary log/rethrow - it will be obvious from the stacktrace that the exception happened in `initiateWrite`.